[quant][core][feature] Implemented masked_fill for QuantizedCPU tensors

dzdang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #78384
* __->__ #78368

Summary:
Previously, there was no suport for masked_fill for quantized tensors.
This PR introduces the feature for quantized CPU tensors. A
corresponding test case was added to test the impl.

Most of the implementation is copied over from the fp kernel and adapted
for quantized tensors.

TODO: implement masked_fill for QuantizedCUDA tensors.

Test Plan:
```
python test/test_quantization.py -k test_qtensor_masked_fill
```

Differential Revision: [D36706178](https://our.internmc.facebook.com/intern/diff/D36706178)

dzdang committed:
[quant][core][feature] Implemented masked_fill for QuantizedCPU tensors

Summary:
Previously, there was no suport for masked_fill for quantized tensors.
This PR introduces the feature for quantized CPU tensors. A
corresponding test case was added to test the impl.

TODO: implement masked_fill for QuantizedCUDA tensors.

Test Plan:
```
python test/test_quantization.py -k test_qtensor_masked_fill
```

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][feature] Implemented masked_fill for QuantizedCPU tensors"


Summary:
Previously, there was no suport for masked_fill for quantized tensors.
This PR introduces the feature for quantized CPU tensors. A
corresponding test case was added to test the impl.

TODO: implement masked_fill for QuantizedCUDA tensors.

Test Plan:
```
python test/test_quantization.py -k test_qtensor_masked_fill
```

Differential Revision: [D36706178](https://our.internmc.facebook.com/intern/diff/D36706178)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36706178).

dzdang committed:
Update on "[quant][core][feature] Implemented masked_fill for QuantizedCPU tensors"


Summary:
Previously, there was no suport for masked_fill for quantized tensors.
This PR introduces the feature for quantized CPU tensors. A
corresponding test case was added to test the impl.

TODO: implement masked_fill for QuantizedCUDA tensors.

Test Plan:
```
python test/test_quantization.py -k test_qtensor_masked_fill
```

Differential Revision: [D36706178](https://our.internmc.facebook.com/intern/diff/D36706178)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][feature] Implemented masked_fill for QuantizedCPU tensors"


Summary:
Previously, there was no suport for masked_fill for quantized tensors.
This PR introduces the feature for quantized CPU tensors. A
corresponding test case was added to test the impl.

TODO: implement masked_fill for QuantizedCUDA tensors.

Test Plan:
```
python test/test_quantization.py -k test_qtensor_masked_fill
```

Differential Revision: [D36706178](https://our.internmc.facebook.com/intern/diff/D36706178)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][feature] Implemented masked_fill for QuantizedCPU tensors"


Summary:
Previously, there was no suport for masked_fill for quantized tensors.
This PR introduces the feature for quantized CPU tensors. A
corresponding test case was added to test the impl.

TODO: implement masked_fill for QuantizedCUDA tensors.

Test Plan:
```
python test/test_quantization.py -k test_qtensor_masked_fill
```

Differential Revision: [D36706178](https://our.internmc.facebook.com/intern/diff/D36706178)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][feature] Implemented masked_fill for QuantizedCPU tensors"


Summary:
Previously, there was no suport for masked_fill for quantized tensors.
This PR introduces the feature for quantized CPU tensors. A
corresponding test case was added to test the impl.

TODO: implement masked_fill for QuantizedCUDA tensors.

Test Plan:
```
python test/test_quantization.py -k test_qtensor_masked_fill
```

Differential Revision: [D36706178](https://our.internmc.facebook.com/intern/diff/D36706178)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][feature] Implemented masked_fill for QuantizedCPU tensors"


Summary:
Previously, there was no suport for masked_fill for quantized tensors.
This PR introduces the feature for quantized CPU tensors. A
corresponding test case was added to test the impl.

Most of the implementation is copied over from the fp kernel and adapted
for quantized tensors.

TODO: implement masked_fill for QuantizedCUDA tensors.

Test Plan:
```
python test/test_quantization.py -k test_qtensor_masked_fill
```

Differential Revision: [D36706178](https://our.internmc.facebook.com/intern/diff/D36706178)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36706178).

dzdang committed:
Update on "[quant][core][feature] Implemented masked_fill for QuantizedCPU tensors"


Summary:
Previously, there was no suport for masked_fill for quantized tensors.
This PR introduces the feature for quantized CPU tensors. A
corresponding test case was added to test the impl.

Most of the implementation is copied over from the fp kernel and adapted
for quantized tensors.

TODO: implement masked_fill for QuantizedCUDA tensors.

Test Plan:
```
python test/test_quantization.py -k test_qtensor_masked_fill
```

Differential Revision: [D36706178](https://our.internmc.facebook.com/intern/diff/D36706178)

[ghstack-poisoned]

vkuzo commented on aten/src/ATen/native/TensorAdvancedIndexingUtils.h:
```diff
@@ -0,0 +1,93 @@
+/*
+The contents of this file are used in ATen/native/TensorAdvancedIndexing.cpp and ATen/native/quantized/TensorAdvancedIndexing.cpp
```
nit: I'd remove this, it's easy for this to become outdated

dzdang replied:
will do

vkuzo replied:
would be good to resolve before landing

dzdang commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -3597,10 +3598,39 @@ void dequantize_tensor_per_tensor_affine_sub_byte_cpu(
+
+void masked_fill_kernel_quantized_cpu(TensorIterator& iter, const Scalar& value, double scale, int zero_point) {
+  AT_DISPATCH_QINT_TYPES(iter.dtype(), "masked_fill", [&] {
+    float scalar_val = value.to<float>();
```
good catch! I'll make that change

dzdang committed:
Update on "[quant][core][feature] Implemented masked_fill for QuantizedCPU tensors"


Summary:
Previously, there was no suport for masked_fill for quantized tensors.
This PR introduces the feature for quantized CPU tensors. A
corresponding test case was added to test the impl.

Most of the implementation is copied over from the fp kernel and adapted
for quantized tensors.

TODO: implement masked_fill for QuantizedCUDA tensors.

Test Plan:
```
python test/test_quantization.py -k test_qtensor_masked_fill
```

Differential Revision: [D36706178](https://our.internmc.facebook.com/intern/diff/D36706178)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36706178).

dzdang committed:
Update on "[quant][core][feature] Implemented masked_fill for QuantizedCPU tensors"


Summary:
Previously, there was no suport for masked_fill for quantized tensors.
This PR introduces the feature for quantized CPU tensors. A
corresponding test case was added to test the impl.

Most of the implementation is copied over from the fp kernel and adapted
for quantized tensors.

TODO: implement masked_fill for QuantizedCUDA tensors.

Test Plan:
```
python test/test_quantization.py -k test_qtensor_masked_fill
```

Differential Revision: [D36706178](https://our.internmc.facebook.com/intern/diff/D36706178)

[ghstack-poisoned]
